### ISSUE TITLE
Using colmap for camera extrinsics calibration

### ISSUE THREAD

**Issue by wf34** [colmap/colmap#1398]
I aim to create a calibration tool, which does not rely on any calibration objects.
A plan is to align localization trajectories from different sensors. I use colmap for a camera.
I've described my vision in the great detail [here](https://robotics.stackexchange.com/questions/23138/find-a-mistake-in-transformation-between-different-coordinate-frames-in-the-wip).

The issue is that I see a constrant, unaccounted for, term in localizations. Which makes calibration impossible.
I thought maybe, any member of colmap commutity would care to comment, here or there.

Thank you, guys, for your time.

**Comment by tsattler** [comment]
I am not sure what you mean by "a constrant, unaccounted for, term in localizations". Are you accounting for the fact that any model estimated by Structure-from-Motion software such as Colmap is only defined up to an arbitrary rotation, arbitrary translation, and arbitrary scaling (in other words, one unit in the 3D model coordinate system has no physical meaning)?

**Comment by wf34** [comment]
>Are you accounting for the fact that any model estimated by Structure-from-Motion software such as Colmap is only defined up to an arbitrary rotation, arbitrary translation, and arbitrary scaling 

Most certainly.
I localize two sensors (lidar and camera, bolted to one agent) in two localization maps, which are consistent, and are in the same world coordinate frame.
 
 **Mapping** 
At a time of preparing these localization maps, (I describe here in very broad strokes) mapping agent does the lidar slam, which yields the lidar localization map. I use agents poses at times of lidar scans to interpolate to camera frame timestamps.
The mapping agent poses at camera times are used in colmap as a prior. I hold poses _fixed_ in colmap optimization, I do only kpt extraction, descriptor matching and triangulation of points. So, maps are consistent. The scale shouldn't be issue also, because we fix camera poses. The colmap localization map is in a camera coordinate frame. `frame pose in colmap = lidar_from_camera_extrinsics.inverse() * agents interpolated pose from lidar slam * lidar_from_camera_extrinsics`. We assume here that the extrinsics of mapping agent were well calibrated.

**Calibration**
the other, calibrating agent uses its lidar and camera to localize in these two maps, and builds two trajectories. 
I expect the discrepancy between two trajectories to be bell-shaped, and zero-mean, if calibrating agent had correct extrinsincs.
But this does not hold, I see constant bias. 
However, when `calibrating agent == mapping agent, agents hardware and extrinics are unchanged between mapping and calibration attempt`, only then the system meets my expectations.

**Comment by tsattler** [comment]
When you talk about the discrepancy between both trajectories, then this takes the offset between both sensors into account?

**Comment by wf34** [comment]
>When you talk about the discrepancy between both trajectories, then this takes the offset between both sensors into account?
Yes, the offset is taken into account.

I've  talked about applying extrinsics in the map building above:
> The colmap localization map is in a camera coordinate frame. frame pose in colmap = lidar_from_camera_extrinsics.inverse() * agents interpolated pose from lidar slam * lidar_from_camera_extrinsics.

I do similarly (backwards, actually) with the camera localization trajectory, to bring it back to the lidar coordinate frame.

**Comment by wf34** [comment]
I've been making some progress with my problem. I've got incremental improvements with the following steps:
- I've moved from the `2`-agent setup to the `N`-agent setup. I hold out `i`-th agent, build the colmap localization map with `N-1` tracks from all agents other than `i`-th. I see less biased localization on the map like that.
- Another change is Global Bundle Adjustment. Previously I've only done colmap's `Points Triangulation` for the localization map building (agent poses were fixed in map).  This change, paradoxically enough, makes the map both better and worth.
BA-map is better in a sense of having crisper edges, thinner walls - a definite feature of convergence. However, BA-map has the agents poses moved in the manner, defying physics: e.g., the BA procedure have shifted everything a few meters to the side. So, as result, I get unphysical sensor calibrations: it's translation part has norm about `3m`, which is bigger than the whole agent.
- So, my next step is to look up how to do Global Bundle Adjustment in colmap with the strong pose priors.
